### ISSUE TITLE
Improvement idea

### ISSUE THREAD

**Issue by MCJOHN974** [yegor256/micromap#207]
I didn't implement and benchmark this, but it potentially could be interesting:

Let's have two arrays instead of one:
``` rust
    pairs0: [MaybeUninit<(K, V)>; N],    // for keys whose first bit is 0
    pairs1: [MaybeUninit<(K, V)>; N],    // for keys whose first bit is 1
```
The idea under them is pretty simple -- in the first array we will store pairs, where first (or any other fixed) bit is 0, the rest -- in the other. 

My intuition why it would work:
- extracting first bit sounds as relatively cheap operations, something like just one BITAND and one CMP native instructions, while it can save as much as `N / 2` iterations in some cases.
- even if it don't help, and all data unfortunately went to same bucket, it generates much smaller overhead, comparing to profit it makes in positive scenarios.
- probably taking first bit of structure is not so bad "hash function" and it will be relatively a lot of positive scenarios (for example if key is an integer, which is relatively often used as a key, it should pretty good).

Some thoughts about good implementation:
- Don't keep two arrays in reality, just assume that numbers with first bit 0 are in first half of the array
- Maybe it will worth to use this logic only for big micromaps (size 128, 64, probably 32)

Did someone tested such approach? Maybe any analogues you benchmarking against?

**Comment by owtotwo** [comment]
This is an interesting optimization approach. However, overall, I am not optimistic about this attempt. 🤔

1. The goal of this project is to create a very small data structure, as stated in the [README](https://github.com/yegor256/micromap/blob/1e2a14177b480a206ca7605bee7f07d07b9b3d7a/README.md?plain=1#L20):
> "It was observed that when a map contains more than 20 keys, it may be better to use the standard [HashMap](https://doc.rust-lang.org/std/collections/struct.HashMap.html), since the performance of micromap::Map may start to degrade."
2. Moreover, this data structure is primarily used on the stack, where space is relatively expensive, so doubling the size of the data structure has a significant impact.
3. Additionally, when the array is small enough, due to memory locality principles, iteration efficiency can be very high if cache hit rates are stable.
4. For arbitrary key-value pairs with unknown structure, taking any bit from them cannot deterministically ensure an equal distribution according to probability theory, so using two arrays does not guarantee efficiency. For example, for a Unix timestamp key, the least significant bit has better uniformity compared to the highest bit which is always zero.
5. Moreover, even selecting the lowest bit of the key as a "simple hash value" is unsuitable because we cannot assume that users' keys will have such properties(the specific bits are not almost always 0 or 1), leading to unstable performance.
6. Such optimizations are typically considered only under sufficiently large constants, meaning this could be a difficult-to-successful space-for-time operation.

(The above content is based on my experience implementing an double-array-based lock-free structure where I used the highest bit as a swap indicator for the double array and tested its performance. The optimization you mentioned is somewhat similar.)

In summary, for small arrays, most optimizations are likely to have negative effects (except common optimizations like `remove_swap` instead of `empty slots`) because CPU Cache Line efficiency is extremely high.

Finally, I haven't done any actual coding or testing, so I can't prove for sure that this optimization is invalid. If you think it might work, you can try coding and testing it, which I think won't be very complicated. 😄

If anyone has a better idea, I hope you can propose it. It may be an effective optimization that we can't think of, which will be very helpful to this project. (If it is a brand new idea, you can open another issue for discussion. If we have no objection to the optimization ideas of this issue, I think we can also consider closing this issue.)

**Comment by avl** [comment]
I think this should be closed as clearly outside the scope of a project named "micromap".
